src/sync/locks.rs

```diff
@@ -0,0 +1,198 @@
+  pub fn try_lock(&self) -> Option<MutexGuard<'_, T>> {
+    if self.raw.raw_lock() {
+      Some(MutexGuard { underlying: self, ptr: self.data.get() })
+    } else {
+      None
+    }
+  }
+}
+unsafe impl<T> Send for Mutex<T> {}
+unsafe impl<T> Sync for Mutex<T> {}
+
+/// A guard representing an active lock on an [`Mutex`].
+pub struct MutexGuard<'a, T> {
+  underlying: &'a Mutex<T>,
+  ptr: *mut T,
+}
+impl<'a, T> Drop for MutexGuard<'a, T> {
+  fn drop(&mut self) {
+    self.underlying.raw.raw_unlock();
+  }
+}
+impl<'a, T> Deref for MutexGuard<'a, T> {
+  type Target = T;
+  fn deref(&self) -> &Self::Target {
+    unsafe { &*self.ptr }
+  }
+}
+impl<'a, T> DerefMut for MutexGuard<'a, T> {
+  fn deref_mut(&mut self) -> &mut Self::Target {
+    unsafe { &mut *self.ptr }
+  }
+}
+
+enum Void {}
+
+/// A helper type that ensures a particular value is only initialized once.
+pub struct InitOnce<T> {
+  state: Static<u8>,
+  value: UnsafeCell<MaybeUninit<T>>,
+}
+impl<T> InitOnce<T> {
+  /// Creates a new uninitialized object.
+  pub const fn new() -> Self {
+    InitOnce { state: Static::new(0), value: UnsafeCell::new(MaybeUninit::uninit()) }
+  }
+
+  /// Gets the contents of this state, or initializes it if it has not already
+  /// been initialized.
+  ///
+  /// The initializer function is guaranteed to only be called once.
+  ///
+  /// Take care when sharing an `InitOnce` object between an IRQ and normal
+  /// code. If this function is called in an IRQ when it is already currently
+  /// being initialized by user code, this function will panic.
+  pub fn get(&self, initializer: impl FnOnce() -> T) -> &T {
+    match self.try_get(|| -> Result<T, Void> { Ok(initializer()) }) {
+      Ok(v) => v,
+      _ => unimplemented!(),
+    }
+  }
+
+  /// Gets the contents of this state, or initializes it if it has not already
+  /// been initialized.
+  ///
+  /// The initializer function is guaranteed to only be called once if it
+  /// returns `Ok`. If it returns `Err`, it will be called again in the
+  /// future until an attempt at initialization succeeds.
+  ///
+  /// Take care when sharing an `InitOnce` object between an IRQ and normal
+  /// code. If this function is called in an IRQ when it is already currently
+  /// being initialized by user code, this function will panic.
+  pub fn try_get<E>(&self, initializer: impl FnOnce() -> Result<T, E>) -> Result<&T, E> {
+    unsafe {
+      if self.state.read() != 2 {
+        // Locks the initializer
+        if self.state.replace(1) != 0 {
+          panic!("Attempt to initialize `InitOnce` that is already in initialization.");
+        }
+
+        // Initialize the actual value.
+        let init = match initializer() {
+          Ok(v) => v,
+          Err(e) => {
+            assert_eq!(self.state.replace(0), 1);
+            return Err(e);
+          }
+        };
+        ptr::write_volatile((*self.value.get()).as_mut_ptr(), init);
+        assert_eq!(self.state.replace(2), 1);
```

Comment:
You need a release fence between writing to `self.value.get()` and writing 2 to `self.state`.

Reply:
Added. Unfortunate with the Rust bug with `__sync_synchronize`, but I guess there's not much to do with that except wait for it to be fixed. x-x